### PR TITLE
Merge Climber Branch to master

### LINKED ISSUE #24: [Programming]Climber code

### PR DESCRIPTION
**Pull request by nick-mcgill**

### CODE REVIEW

File: src/main/cpp/ControllerInterface.cpp [hunk 1]
```diff
@@ -54,4 +55,10 @@ void ControllerInterface::UpdateSwerveInput(RobotControlData &controlData)
         controlData.swerveInput.yTranslation *= m_slowmodefactor;
         controlData.swerveInput.rotation *= m_slowmodefactor;
     }
-};
\ No newline at end of file
+};
+
+void ControllerInterface::UpdateClimberInput(RobotControlData &controlData)
+{
+    controlData.climberInput.controllerSpeed = m_pilot.GetLeftY();
```
Comment (Harsh-Sinha) [inline src/main/cpp/ControllerInterface.cpp]: make this right y

File: src/main/include/subsystems/SubSystemConfig.h [hunk 2]
```diff
@@ -72,4 +72,12 @@ const double LAUNCHER_ZERO_OFFSET = 13.501;
 const double MAX_PIVOT_ANGLE = 75.0;
 const double MIN_PIVOT_ANGLE = 5.0;
 
-const bool INVERT_INDEXER = false;
\ No newline at end of file
+const bool INVERT_INDEXER = false;
+// ###########################################################
+// #                         Climber                         #
+// ###########################################################
+
```
Comment (Harsh-Sinha) [inline src/main/include/subsystems/SubSystemConfig.h:88]: add a flag to invert the main motor (i.e. the left climber motor)

File: src/main/include/subsystems/subsystemsManagers/ClimberManager.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{
+public:
+    void HandleInput(ClimberInput &input);
+
+    
+private:
+    ClimberHAL m_climber;
+    double m_deadzone = 0.05;
+    double m_multiplier = 0.8;
+    double m_maxHeight = 3.8;
```
Comment (Harsh-Sinha) [inline src/main/include/subsystems/subsystemsManagers/ClimberManager.h]: don't define max height here

File: src/main/include/subsystems/ClimberHAL.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        rev::CANSparkMax m_climberMotorLeft = rev::CANSparkMax(CLIMBER_MOTOR_LEFT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+        rev::CANSparkMax m_climberMotorRight = rev::CANSparkMax(CLIMBER_MOTOR_RIGHT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+
+        rev::SparkRelativeEncoder m_EncoderRight = m_climberMotorRight.GetEncoder(rev::SparkRelativeEncoder::Type::kHallSensor, 1);
+        rev::SparkRelativeEncoder m_EncoderLeft = m_climberMotorLeft.GetEncoder(rev::SparkRelativeEncoder::Type::kHallSensor, 1);
+    
+        rev::REVLibError error;
+
+        //Constants
+        const double wheel_radius = 1.5;
```
Comment (Harsh-Sinha) [inline src/main/include/subsystems/ClimberHAL.h]: for all "double" calculations add a ".0" to the end of the numbers

File: src/main/include/subsystems/ClimberHAL.h [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        const double wheel_circumference = 2 * 3.14 * wheel_radius;
+        const double maxHeight = (3 + 8/12) * 12; //inches
```
Comment (Harsh-Sinha) [inline src/main/include/subsystems/ClimberHAL.h]: const variables should be all caps... see subsystem config for example

File: src/main/include/subsystems/ClimberHAL.h [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+class ClimberHAL{
+    public:
+        ClimberHAL();
+        void RunClimber(double controllerSpeed);
+        bool isNextRotationTooHigh(double height);
+    private:
+        rev::CANSparkMax m_climberMotorLeft = rev::CANSparkMax(CLIMBER_MOTOR_LEFT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+        rev::CANSparkMax m_climberMotorRight = rev::CANSparkMax(CLIMBER_MOTOR_RIGHT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+
+        rev::SparkRelativeEncoder m_EncoderRight = m_climberMotorRight.GetEncoder(rev::SparkRelativeEncoder::Type::kHallSensor, 1);
```
Comment (Harsh-Sinha) [inline src/main/include/subsystems/ClimberHAL.h]: number should be 42, not 1

File: src/main/cpp/subsystems/ClimberHAL.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#include "subsystems/ClimberHAL.h"
+
+ClimberHAL::ClimberHAL() {
+    m_climberMotorLeft.RestoreFactoryDefaults();
+    m_climberMotorRight.RestoreFactoryDefaults();
+    
+    error = m_EncoderLeft.SetPositionConversionFactor(wheel_circumference); 
+
```
Comment (Harsh-Sinha) [inline src/main/cpp/subsystems/ClimberHAL.cpp:9]: invert left motor based on flag defined in sub system config

**Review by Harsh-Sinha — CHANGES_REQUESTED** [review]

File: src/main/include/subsystems/ClimberHAL.h (hunk 6 above)
Comment (BlueStrawberryLY) [reply src/main/include/subsystems/ClimberHAL.h]: done

File: src/main/cpp/ControllerInterface.cpp (hunk 1 above)
Comment (hluka) [reply src/main/cpp/ControllerInterface.cpp]: Done

File: src/main/include/subsystems/ClimberHAL.h (hunk 5 above)
Comment (BlueStrawberryLY) [reply src/main/include/subsystems/ClimberHAL.h]: Done

File: src/main/include/subsystems/subsystemsManagers/ClimberManager.h (hunk 3 above)
Comment (BlueStrawberryLY) [reply src/main/include/subsystems/subsystemsManagers/ClimberManager.h]: done

File: src/main/include/subsystems/ClimberHAL.h (hunk 4 above)
Comment (BlueStrawberryLY) [reply src/main/include/subsystems/ClimberHAL.h]: done

File: src/main/cpp/subsystems/ClimberHAL.cpp (hunk 7 above)
Comment (JoeyBrar) [reply src/main/cpp/subsystems/ClimberHAL.cpp:9]: done

File: src/main/include/subsystems/SubSystemConfig.h (hunk 2 above)
Comment (JoeyBrar) [reply src/main/include/subsystems/SubSystemConfig.h:88]: done

File: src/main/include/subsystems/ClimberHAL.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    public:
+        ClimberHAL();
+        void RunClimber(double controllerSpeed);
+        bool isNextRotationTooHigh(double height);
+    private:
+        rev::CANSparkMax m_climberMotorLeft = rev::CANSparkMax(CLIMBER_MOTOR_LEFT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+        rev::CANSparkMax m_climberMotorRight = rev::CANSparkMax(CLIMBER_MOTOR_RIGHT, rev::CANSparkLowLevel::MotorType::kBrushless); 
+
+        rev::SparkRelativeEncoder m_EncoderRight = m_climberMotorRight.GetEncoder(rev::SparkRelativeEncoder::Type::kHallSensor, 42);
+        rev::SparkRelativeEncoder m_EncoderLeft = m_climberMotorLeft.GetEncoder(rev::SparkRelativeEncoder::Type::kHallSensor, 1);
```
Comment (nick-mcgill) [inline src/main/include/subsystems/ClimberHAL.h]: Does this one also need to be 42?  Could we create a constant for this?

File: src/main/include/subsystems/subsystemsManagers/ClimberManager.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+class ClimberManager
+{
+public:
+    void HandleInput(ClimberInput &input);
+
+    
+private:
+    ClimberHAL m_climber;
+    double m_deadzone = 0.05;
```
Comment (nick-mcgill) [inline src/main/include/subsystems/subsystemsManagers/ClimberManager.h]: Should these be defined as constants?

**Comment by nick-mcgill** [comment]
@Rvistix @Harsh-Sinha Merge conflicts resolved and branch available to merge to master whenever you're ready.

**Pull request closed without merging**